agatha/construct/semrep_util.py

```diff
@@ -321,15 +423,28 @@ def run(self, input_path:Path, output_path:Path)->None:
     input_path = Path(input_path)
     assert input_path.is_file(), f"Failed to find {input_path}"
     assert not output_path.exists(), f"Refusing to overwrite {output_path}"
-    cmd = self._get_flags(input_path, output_path)
-    env = self._get_env()
-    print("Running:", " ".join(cmd))
-    subprocess.run(
-        cmd,
-        env=env,
-        check=True,
-    )
-    assert output_path.is_file(), f"SemRep Failed to produce {output_path}"
+    if line_mode:
+      buffer = []
+      lines = self._read_lines(input_path)
+      for line in lines:
+        try:
+          buffer += self._run_gracefully([line])
```

Comment:
I have bad news for you. This is what a very simple call to semrep will output.

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE SemRepAnnotation PUBLIC "-//NLM//DTD SemRep Output//EN" "http://semrep.nlm.nih.gov/DTD/SemRepXML_v1.7.dtd">
<SemRepAnnotation>
  <Document id="D00000000" text="Tobacco causes cancer">
    <Utterance id="D00000000.tx.1" section="tx" number="1" text="Tobacco causes cancer">
      <Entity id="D00000000.E1" cui="C0040329" name="Tobacco" semtypes="hops,orch,phsu" text="Tobacco" score="1000" begin="0" end="7"/>
      <Entity id="D00000000.E2" cui="C0006826" name="Malignant Neoplasms" semtypes="neop" text="cancer" score="1000" begin="15" end="21"/>
      <Predication id="D00000000.P1">
        <Subject maxDist="1" dist="1" entityID="D00000000.E1" relSemType="hops"/>
        <Predicate type="CAUSES" indicatorType="VERB" begin="8" end="14"/>
        <Object maxDist="1" dist="1" entityID="D00000000.E2" relSemType="neop"/>
      </Predication>
    </Utterance>
  </Document>
</SemRepAnnotation>
```

That header:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE SemRepAnnotation PUBLIC "-//NLM//DTD SemRep Output//EN" "http://semrep.nlm.nih.gov/DTD/SemRepXML_v1.7.dtd">
```

Is going to occur, _even_ if the process breaks later on.

This means that the xml parser is going to fail if it find a file with multiple headers scattered through the file.

Reply:
I think what you can do instead is take advantage of `StringIO`
https://docs.python.org/3/library/io.html

This way, when you read in a very short xml file from semrep, you can pass that data directly to the `lxml` parser without needed the files.

This ultimately means that we're going to have to change the interface a little. Now, `run` will probably have to return the records produced by `def _parse_semrep_xml_entity(xml_entity:lxml.etree._Element)->Record:` directly.

Reply:
I'm not sure I understand. We want to return `Record` from the `run()` or we want to return `lxml.etree._Element`?